marklogic-client-api/src/main/java/com/marklogic/client/datamovement/ProgressListener.java

```diff
@@ -0,0 +1,189 @@
+
+	/**
+	 * Use this constructor for when the total number of results isn't known ahead of time.
+	 *
+	 * @param consumers
+	 */
+	public ProgressListener(Consumer<ProgressUpdate>... consumers) {
+		this(0, consumers);
+	}
+
+	/**
+	 * Use this constructor for when the total number of results is known ahead of time. The ProgressUpdate objects that
+	 * are emitted will have more interesting information as a result.
+	 *
+	 * @param consumers
+	 * @param totalResults
+	 */
+	public ProgressListener(long totalResults, Consumer<ProgressUpdate>... consumers) {
+		this.totalResults = totalResults;
+		for (Consumer<ProgressUpdate> consumer : consumers) {
+			this.consumers.add(consumer);
+		}
+	}
+
+	/**
+	 * Initializes the start time so that each ProgressUpdate knows how long it occurred after the job was started.
+	 *
+	 * @param queryBatcher
+	 */
+	@Override
+	public void initializeListener(QueryBatcher queryBatcher) {
+		startTime = System.currentTimeMillis();
+	}
+
+	/**
+	 * Batches arrive in random order, so a ProgressUpdate is created and sent to each Consumer only if the
+	 * value of "getJobResultsSoFar" on the QueryBatch exceeds the number of results seen so far.
+	 * <p>
+	 * For example, if there are 2 batches, and batch 2 is processed first by this listener followed by batch 1, a
+	 * ProgressUpdate is only created when batch 2 is processed.
+	 *
+	 * @param batch
+	 */
+	@Override
+	public void processEvent(QueryBatch batch) {
+		// resultsSoFar is an AtomicLong so it can be safely updated across many threads
+		final long jobResultsSoFar = batch.getJobResultsSoFar();
+		final long newResultsSoFar = this.resultsSoFar.updateAndGet(operand ->
+			jobResultsSoFar > operand ? jobResultsSoFar : operand
+		);
+		boolean resultsSoFarWasUpdated = jobResultsSoFar == newResultsSoFar;
+
+		if (resultsSoFarWasUpdated && consumers != null) {
+			double timeSoFar = ((double) System.currentTimeMillis() - startTime) / 1000;
+
+			/**
+			 * The initial totalResults may have been incorrectly set to a value lower than jobResultsSoFar; if this
+			 * occurs, use jobResultsSoFar as the value passed to the ProgressUpdate object. totalResults is not
+			 * updated though in case there's a need to know what the initial value was.
+			 */
+			long totalForThisUpdate = jobResultsSoFar > this.totalResults && this.totalResults > 0 ? jobResultsSoFar : this.totalResults;
+
+			ProgressUpdate progressUpdate = new SimpleProgressUpdate(batch, startTime, totalForThisUpdate, timeSoFar);
+
+			for (Consumer<ProgressUpdate> consumer : consumers) {
+				invokeConsumer(consumer, progressUpdate);
+			}
+		}
+	}
+
+	/**
+	 * Protected so that a subclass can override how a consumer is invoked, particularly how an exception is handled.
+	 *
+	 * @param consumer
+	 * @param progressUpdate
+	 */
+	protected void invokeConsumer(Consumer<ProgressUpdate> consumer, ProgressUpdate progressUpdate) {
+		try {
+			consumer.accept(progressUpdate);
+		} catch (Throwable t) {
```

Comment:
Normally I'd want to distinguish more between OK errors and ones that need to be rethrown.  It might help somebody testing and debugging later on  if Throwable is refined a bit.  Up to you.

Reply:
I followed what ExportListener is doing as it iterates over consumers - it catches a Throwable and logs it as well.